[receiver/datadog] Support 128 bits TraceID coming from Datadog

xiu opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
With this commit, we add support for 128 bits TraceIDs coming from Datadog instrumented services. This can happen when an OTel instrumented service calls a downstream Datadog instrumented one. Datadog instrumentation libraries store the 128 bits TraceID into two different fields:
* TraceID: lower 64 bits of the 128 bits TraceID 
* _dd.p.tid: upper 64 bits of the 128 bits TraceID

This commit adds logic that reconstructs the 128 bits TraceID. Before this commit, only the lower 64 bits were used as TraceID.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #36926

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Tested the setup with the following chain: OTel Instrumented Service --> Datadog Instrumented Service -> OTel Instrumented Service. The TraceID was maintained on the whole chain.

Also added a unit test (`TestToTraces64to128bits`).

#### Documentation
Updated README.md in [58129d5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39654/commits/58129d562a1302e849e95f9a4ad98446e35cc5a1)

xiu committed:
[receiver/datadog] Support 128 bits TraceID coming from Datadog

With this commit, we add support for 128 bits TraceIDs coming from Datadog
instrumented services. This can happen when an OTel instrumented service calls
a downstream Datadog instrumented one. Datadog instrumentation libraries store
the 128 bits TraceID into two different fields:
* TraceID: lower 64 bits of the 128 bits TraceID
* _dd.p.tid: upper 64 bits of the 128 bits TraceID

This commit adds logic that reconstructs the 128 bits TraceID. Before this
commit, only the lower 64 bits were used as TraceID.

Fixes #36926

dehaansa commented on receiver/datadogreceiver/internal/translator/traces_translator.go:
```diff
@@ -90,6 +90,13 @@ func ToTraces(payload *pb.TracerPayload, req *http.Request) ptrace.Traces {
+	// to concatenate the two into newSpan.TraceID.
+	// The following map keeps track of the TraceIDs we process as only the first span has the upper 64 bits from the
+	// 128 bits trace ID.
+	map64to128 := make(map[uint64]pcommon.TraceID)
```
I'm concerned about an unbounded cache. While each entry is small, this is just unbounded memory growth.

It looks like the `spanmetricsconnector` uses a LRU (least recently used) cache from hashicorp `"github.com/hashicorp/golang-lru/v2/simplelru"` that would be a good candidate to use here to ensure there is a limit to the size of this cache.

xiu replied:
This is a good point, the cache would just be bound to that one `ToTraces` run, but I'm not sure what a good size for it would be. Should it be configurable? If not, what would a good value be?

xiu replied:
Actually it seems fairly easy to add such a config, I'll add that.

xiu replied:
Done in [90eb40c](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39654/commits/90eb40ca0f6cbf2cb734dfbe863e78f0b97a6387)

dehaansa reviewed (commented):
Thanks for taking the time to dig in to this issue!

This seems like a very useful fix, but would probably constitute a breaking change for existing users of the component. As such I think we'll need to make it opt-in using a featuregate.

You can check for examples of using featuregates in other components, but they're relatively easy to set up. If you have one set to StageAlpha, it will be disabled by default, and then in later versions we will move it to Beta (enabled by default) and then the gate can later be removed. The gate will then need to be configured at runtime using a flag like `--feature-gates=receiver.datadogreceiver.Enable128BitTraceID`

Here's an example of what defining the gate looks like in code, copied and modified slightly from another component

```
var FullTraceIDFeatureGate = featuregate.GlobalRegistry().MustRegister(
	"receiver.datadogreceiver.Enable128BitTraceID",
	featuregate.StageAlpha,
	featuregate.WithRegisterDescription("<description."),
	featuregate.WithRegisterFromVersion("v0.125.0"),
)

...

if FullTraceIDFeatureGate.Enabled() {
...

xiu committed:
Change TraceID cache to LRU and move it to the receiver

This helps with reusing trace id across multiple flushes. This commit also adds
a new configuration item: `trace_id_cache_size`, which will be used to size the
LRU cache.

If translator.ToTraces doesn't get a cache (e.g. nil), it'll disable the
feature.

xiu committed:
Add featuregate `receiver.datadogreceiver.Enable128BitTraceID`

xiu commented:
> Thanks for taking the time to dig in to this issue!
> 
> This seems like a very useful fix, but would probably constitute a breaking change for existing users of the component. As such I think we'll need to make it opt-in using a featuregate.
> 
> You can check for examples of using featuregates in other components, but they're relatively easy to set up. If you have one set to StageAlpha, it will be disabled by default, and then in later versions we will move it to Beta (enabled by default) and then the gate can later be removed. The gate will then need to be configured at runtime using a flag like `--feature-gates=receiver.datadogreceiver.Enable128BitTraceID`
> 
> Here's an example of what defining the gate looks like in code, copied and modified slightly from another component
> 
> ```
> var FullTraceIDFeatureGate = featuregate.GlobalRegistry().MustRegister(
> 	"receiver.datadogreceiver.Enable128BitTraceID",
> 	featuregate.StageAlpha,
> 	featuregate.WithRegisterDescription("<description."),
> 	featuregate.WithRegisterFromVersion("v0.125.0"),
> )
> 
> ...
> 
> if FullTraceIDFeatureGate.Enabled() {
> ...
> ```

@dehaansa Thanks for reviewing! 

I've added a featuregate in [af83cf2](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39654/commits/af83cf2a47688423e45f81c683413fc6e3c9fe68), let me know if anything.

xiu committed:
Update README.md

dehaansa reviewed (approved):
LGTM, will try to get codeowner review before marking ready for merge.

xiu committed:
make gotidy

cyrille-leclerc commented:
When testing activating `--feature-gates=receiver.datadogreceiver.Enable128BitTraceID`, I get the following failure. I'll capture the trace id to better understand:

```
2025-05-03T17:17:06.934+0200    error   datadogreceiver@v0.125.0/receiver.go:254        Error converting traces {"otelcol.component.id": "datadog", "otelcol.component.kind": "receiver", "otelcol.signal": "traces", "error": "hex encoded trace-id must have length equals to 32"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver.(*datadogReceiver).handleTraces
        github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver@v0.125.0/receiver.go:254
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
net/http.(*ServeMux).ServeHTTP
        net/http/server.go:2822
go.opentelemetry.io/collector/config/confighttp.(*decompressor).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/compression.go:183
go.opentelemetry.io/collector/config/confighttp.(*ServerConfig).ToServer.maxRequestBodySizeInterceptor.func2
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/confighttp.go:615
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:179
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:67
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/collector/config/confighttp.(*clientInfoHandler).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/clientinfohandler.go:26
net/http.serverHandler.ServeHTTP
        net/http/server.go:3301
net/http.(*conn).serve
        net/http/server.go:2102
2025-05-03T17:17:09.015+0200    error   datadogreceiver@v0.125.0/receiver.go:254        Error converting traces {"otelcol.component.id": "datadog", "otelcol.component.kind": "receiver", "otelcol.signal": "traces", "error": "hex encoded trace-id must have length equals to 32"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver.(*datadogReceiver).handleTraces
        github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver@v0.125.0/receiver.go:254
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
net/http.(*ServeMux).ServeHTTP
        net/http/server.go:2822
go.opentelemetry.io/collector/config/confighttp.(*decompressor).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/compression.go:183
go.opentelemetry.io/collector/config/confighttp.(*ServerConfig).ToServer.maxRequestBodySizeInterceptor.func2
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/confighttp.go:615
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:179
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:67
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/collector/config/confighttp.(*clientInfoHandler).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/clientinfohandler.go:26
net/http.serverHandler.ServeHTTP
        net/http/server.go:3301
net/http.(*conn).serve
        net/http/server.go:2102
^C2025-05-03T17:17:09.866+0200  info    otelcol@v0.125.1-0.20250430101735-2ecd0b7f95cd/collector.go:358 Received signal from OS {"signal": "interrupt"}
2025-05-03T17:17:09.866+0200    info    service@v0.125.1-0.20250430101735-2ecd0b7f95cd/service.go:331   Starting shutdown...
2025-05-03T17:17:09.866+0200    info    healthcheck/handler.go:132      Health Check state change       {"otelcol.component.id": "health_check", "otelcol.component.kind": "extension", "status": "unavailable"}
2025-05-03T17:17:09.867+0200    info    spanmetricsconnector@v0.125.0/connector.go:219  Shutting down spanmetrics connector     {"otelcol.component.id": "spanmetrics", "otelcol.component.kind": "connector", "otelcol.signal": "traces", "otelcol.signal.output": "metrics"}
2025-05-03T17:17:09.867+0200    info    spanmetricsconnector@v0.125.0/connector.go:221  Stopping ticker {"otelcol.component.id": "spanmetrics", "otelcol.component.kind": "connector", "otelcol.signal": "traces", "otelcol.signal.output": "metrics"}
2025-05-03T17:17:09.867+0200    info    grafanacloudconnector@v0.125.0/connector.go:114 Stopping Grafana Cloud connector        {"otelcol.component.id": "grafanacloud", "otelcol.component.kind": "connector", "otelcol.signal": "traces", "otelcol.signal.output": "metrics"}
2025-05-03T17:17:10.485+0200    info    extensions/extensions.go:69     Stopping extensions...
2025-05-03T17:17:10.485+0200    info    service@v0.125.1-0.20250430101735-2ecd0b7f95cd/service.go:345   Shutdown complete.
```

cyrille-leclerc commented on receiver/datadogreceiver/internal/translator/traces_translator.go:
```diff
@@ -102,6 +103,33 @@ func ToTraces(payload *pb.TracerPayload, req *http.Request) ptrace.Traces {
+			if traceIDCache != nil {
+				if val, ok := span.Meta["_dd.p.tid"]; ok {
+					hexTraceID := val + strconv.FormatUint(span.TraceID, 16)
+					traceID, err := oteltrace.TraceIDFromHex(hexTraceID)
```
Shall we enrich the error message with the provided `hexTraceID` to troubleshoot the error `hex encoded trace-id must have length equals to 32`?

xiu replied:
Added

cyrille-leclerc commented on receiver/datadogreceiver/internal/translator/traces_translator.go:
```diff
@@ -195,14 +223,14 @@ func tagsToSpanLinks(tags map[string]string, dest ptrace.SpanLinkSlice) error {
 		// Convert trace id.
-		rawTrace, errTrace := trace.TraceIDFromHex(span.TraceID)
+		rawTrace, errTrace := oteltrace.TraceIDFromHex(span.TraceID)
 		if errTrace != nil {
```
Ditto, shall we enrich the error message with the provided `hexTraceID` to troubleshoot the error `hex encoded trace-id must have length equals to 32`?

xiu replied:
Added a wrapping here and on the call to SpanIDFromHex below as well (not related to this PR). This gets me to wonder whether span links will work with the reconstructed spanIDs, I'll need to test that.

xiu committed:
wrap errors

cyrille-leclerc commented:
```
2025-05-04T14:46:39.095+0200    error   datadogreceiver@v0.125.0/receiver.go:254        Error converting traces {"otelcol.component.id": "datadog", "otelcol.component.kind": "receiver", "otelcol.signal": "traces", "error": "error converting to a 128bit traceid (_dd.p.tid: 681761ae00000000 - span.TraceID: 531758935570545189): hex encoded trace-id must have length equals to 32"}
```

<details>
<summary>OTel Col stack trace</summary>
```
2025-05-04T14:46:39.095+0200    error   datadogreceiver@v0.125.0/receiver.go:254        Error converting traces {"otelcol.component.id": "datadog", "otelcol.component.kind": "receiver", "otelcol.signal": "traces", "error": "error converting to a 128bit traceid (_dd.p.tid: 681761ae00000000 - span.TraceID: 531758935570545189): hex encoded trace-id must have length equals to 32"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver.(*datadogReceiver).handleTraces
        github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver@v0.125.0/receiver.go:254
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
net/http.(*ServeMux).ServeHTTP
        net/http/server.go:2822
go.opentelemetry.io/collector/config/confighttp.(*decompressor).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/compression.go:183
go.opentelemetry.io/collector/config/confighttp.(*ServerConfig).ToServer.maxRequestBodySizeInterceptor.func2
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/confighttp.go:615
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:179
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.60.0/handler.go:67
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2294
go.opentelemetry.io/collector/config/confighttp.(*clientInfoHandler).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.125.1-0.20250430101735-2ecd0b7f95cd/clientinfohandler.go:26
net/http.serverHandler.ServeHTTP
        net/http/server.go:3301
net/http.(*conn).serve
        net/http/server.go:2102
```
</details>

xiu committed:
fix wrong conversion logic and better error handling

As noted by @cyrille-leclerc, some spans failed to convert their trace id to
128 bits. This was due to a naive string concatenation vs properly splitting
the trace id at the byte level as in uInt64ToTraceID. Previously, we were also
rejecting the traces in the current ToTraces run. Now, we just skip converting
and keep the 64-bit trace id if anything.

xiu commented:
Thanks for testing and the help debugging the issue! 
I believe I've fixed the issue with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39654/commits/b3f8d1d25533cce14398f726fbf1d205f9572d1a, it was due to a naive string concatenation rather than working at the byte level through `uInt64ToTraceID`. 

I've also improved error handling so as not to reject the whole `ToTraces` run when an error in encountered but rather keep the 64-bit trace id.

cyrille-leclerc commented:
I could successfully test the fix:
* Upstream service called `frontend` instrumented with otel java auto-instr
* Downstream service `shipping`  instrumented with dd-trace-java emitting telemetry to the otelcol datadog receiver

With otelcol 0.125.0, the traces of the `frontend` service don't connect the `shipping` service which is an error.
With otelcol `--feature-gates=receiver.datadogreceiver.Enable128BitTraceID`, the  the traces of the `frontend` service don't connect the `shipping` service which is the desired behaviour.

xiu committed:
make gci

xiu commented:
Vulnerability addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39852

dehaansa commented:
Could we get a codeowner review? @boostchicken @gouthamve @MovieStoreGuy

MovieStoreGuy commented:
Please resolve the conflicts :)

xiu commented:
@MovieStoreGuy done!
